### DISCUSSION TITLE
Feature Request: Matrix with Zig-Zag supports (32x8 led matrices)

### DISCUSSION THREAD

**Discussion by dandantsui** [PlummersSoftwareLLC/NightDriverStrip#111] [Ideas]
Hello! 

Just wanted to request a feature similar to the Adafruit NeoMatrix library to handle zig-zags on LED matrix strips.

Can such be implemented? Would be a great addition for driving wireless-data from Unity.

[https://learn.adafruit.com/adafruit-neopixel-uberguide/neomatrix-library](https://learn.adafruit.com/adafruit-neopixel-uberguide/neomatrix-library )

[https://www.adafruit.com/product/2294](https://www.adafruit.com/product/2294)

**Comment by MarcJenningsUK** [comment]
A sample implementation of a snaked array processor (in C# - my C++ chops are rusty) would be 

```csharp
public class ArrFunc
{
    public static void Main()
    {
        var inputchars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ".ToCharArray();
 
        var stopwatch = new System.Diagnostics.Stopwatch();
        stopwatch.Start();
        var outchars = Make2DArray(inputchars, 24, 6);
        stopwatch.Stop();
        Console.WriteLine($"Array manipulation took {stopwatch.ElapsedMilliseconds} ms");

        for(int i = 23; i >= 0; i--)
        {
            for (int j = 0; j < 6; j++)
            {
                Console.Write(outchars[i,j] + " ");
            }
            Console.WriteLine();
        }
 
        Console.ReadLine();
    }
 
    static T[,] Make2DArray<T>(T[] input, int height, int width, bool reverse = false)
    {
        T[,] output = new T[height, width];
        for (int i = 0; i < height; i++)
        {
            if(i % 2 == (reverse ? 0 : 1))
            {
                for (int j = 0; j < width; j++)
                {
                    output[i, j] = input[i * width + j];
                }
            }
            else
            {
                for (int j = 0; j < width; j++)
                {
                    output[i, width - 1 - j] = input[i * width + j];
               }
            }
        }
 
        return output;
    } 
 }
```

Apologies for the formatting.  GH hates me.

**Reply by rbergen** [reply]
> Apologies for the formatting. GH hates me.

(No, it doesn't. You just have to use the "three-backtick" multi-line code block markers, and include a language specification where applicable. I've applied it for you in this case, feel free to have a look at it. :) )

**Reply by MarcJenningsUK** [reply]
Thanks.  I just realised, though, that it is probably the inverse of this function that is required.  My bad.

**Reply by rbergen** [reply]
I'm trying to imagine what the "inverse of this function" would be, but I think I'll have to conclude you lost me...

**Reply by MarcJenningsUK** [reply]
I think I just created a method to do the exact opposite of what was required.  It should accept a 2d array and return a 1d array with all the bytes rearranged.

**Reply by rbergen** [reply]
Haha, I thought we were still talking about markdown backticks. 😆 I'll just be quiet from now on.

**Reply by MarcJenningsUK** [reply]
This may be more use, but needs some tweaking for start and end points.

```
    static T[] SnakeToStrip<T>(T[,] input, bool reversed = false)
    {
        int w = input.GetLength(0);
        int h = input.GetLength(1);
        T[] stripValues = new T[w * h];

            for (int i = 0; i < w; i++)
            {
        for (int j = 0; j < h; j++)
        {
                var outlocation = reversed 
                    && j % 2 == 0 ?
                        (j * w) + i : 
                        (j * w) + (w - i - 1);

                stripValues[outlocation] = input[i,j];
            }
        }

        return stripValues;
    }
```
